#### &#128337; Temps previst de dedicació: 10 minuts

# TEMA 1: Casos específics de funcions

## Unitat 3: Definició de funcions dins de funcions

Ja coneixem la sintaxi bàsica que hem de seguir per definir una funció. Però, més enllà d'aquesta forma estàndard, JavaScript ens permet crear funcions amb altres construccions que en alguns casos ens resulten de gran utilitat. Una d'aquestes maneres especials de crear una funció consisteix a fer-ho de manera niada dins d'una altra funció. En aquest cas, la funció niada s'anomena funció interna i la funció que la conté s'anomena funció externa. La funció interna només pot ser utilitzada dins de la funció externa, però no al contrari.

In [1]:
// `finalPrice` is a standard function defined as usual.
function finalPrice(units, unitPrice, coin) {
    const total = units * unitPrice;

    // `coinDescription` is an inner function to be used only from within `finalPrice`.
    function coinDescription(coin) {
        if (coin === '€') {
            return 'euros';
        } else if (coin === '$') {
            return 'dollars';
        } else {
            return 'unknown coin';
        }
    }

    return total.toFixed(2) + ' ' + coinDescription(coin);
}

console.log(finalPrice(30, 2.25, '€'));

67.50 euros


En aquest exemple, la funció `finalPrice()` està definida com hem vist sempre fins a aquest moment. Es tracta d'una funció que calcula el preu final que cal pagar en una compra i, per això, rep tres paràmetres: el nombre de productes comprats, el preu unitari per producte i el símbol de la moneda en la qual es duu a terme la transacció. Alhora, retorna preu final, mostrant l'import amb dos decimals i afegint-hi la descripció de la divisa. Per exemple, si hem comprat 30 unitats a un preu de 2,25, `finalPrice()` ens retorna la cadena de text "67.50 euros".

Per calcular aquest resultat, tanmateix, utilitza una altra funció niada, `coinDescription()`, que resulta accessible només des de l'interior de `finalPrice()`. Aquesta funció interna `coinDescription()` duu a terme una tasca molt específica: convertir un símbol de divisa en la seva descripció. Per exemple, si el símbol de la divisa és "€", `coinDescription()` retorna la cadena de text "euros". Si el símbol de la divisa és "$", `coinDescription()` retorna la cadena de text "dollars". I així successivament amb totes les divises que vulguem incloure.

En realitat podríem transformar la funció `coinDescription()` en una funció ordinària, fora de `finalPrice()` i situada al mateix nivell que aquesta.

In [2]:
// `coinDescription` is a standard function defined as usual.
function coinDescription(coin) {
  if (coin === '€') {
      return 'euros';
  } else if (coin === '$') {
      return 'dollars';
  } else {
      return 'unknown coin';
  }
}

// `totalPrice` is a standard function defined as usual.
function totalPrice(units, unitPrice, coin) {
  const total = units * unitPrice;
  return total.toFixed(2) + ' ' + coinDescription(coin);
}

console.log(totalPrice(30, 2.25, '€'));

67.50 euros


> <br>🚨&nbsp; **Atenció**
> 
> En el codi anterior hem canviat el nom de `finalPrice()` a `totalPrice()` per evitar definir dues vegades una funció amb el mateix nom, pràctica no permesa en JavaScript.
> <br><br>

El resultat final, com podeu observar, és exactament el mateix. Quin criteri hem de seguir, llavors, per decidir si una funció ha de definir-se dins d'una altra funció o no? Es tracta, en bona part, d'una qüestió subjectiva. En el nostre exemple, si tenim la intenció de cridar la funció `coinDescription()` des d'altres parts del codi, sens dubte hauria de ser una funció estàndard, no pas una funció interna. Si, en canvi, només `finalPrice()` i cap altra part del codi necessita utilitzar la funció `coinDescription()`, definir-la com una funció dins de `finalPrice()` pot ajudar-nos a mantenir el nostre codi més ben encapsulat. Convé, això sí, no crear estructures de funcions excessivament niades: amb funcions definides dins d'altres funcions que, alhora, estan definides dins d'altres funcions, etc. Fer-ho d'aquesta manera afegiria una alta dosi de complexitat al nostre codi font.

En altres llenguatges de programació, definir una funció dins d'una altra funció pot resultar complex i no és gens comú. En JavaScript, però, es tracta d'una tècnica senzilla i d'ús habitual entre la comunitat de programadors.